Analyze Past 24 Hours with Beacons Proxy

fullmetalcache opened the pull request:
Shoutout to @lisaSW for closely working together with me on this. Was a team effort!

Previously, beacons proxy was only analyzing the current hour's worth of data. This PR is to address this issue so that beacons proxy will analyze the rolling 24 hour period, as is done with other modules. A new table, uconnproxy, is created to facilitate this behavior.

Beacons proxy no longer uses the proxy IP as part of the key for uconnproxy entries (previously beaconproxymap entries). This is because the proxy IP is not relevant in terms of determining beaconing behavior from a host to an fqdn; the proxy IP purely serves as an intermediary. The uconnproxy entries are now keyed on just src-fqdn pairs.

This PR also includes a bug fix for when the "host" field in the http.log file is blank, which was found to occur in some cases during testing. The backup measure that is used is to parse the fqdn from the URI field. This backup measure is the secondary choice as it takes additional processing time over just pulling the fqdn from the host field.

Minor renaming of mongo variables with respect to the proxy IP was also done to help avoid some confusion with the module (i.e., dst_x info is now "proxy", which contains the relevant host info of ip, network name, network uuid).

The module does not store a list of proxy IPs if the server happens to change during the analysis period; only the most recent proxy server will be stored. 

Closes #670
Closes #691

Zalgo2462 commented on pkg/beaconproxy/mongodb.go:
```diff
@@ -63,7 +63,7 @@ func (r *repo) CreateIndexes() error {
 }
 
```
Please change the indexes line above on lines 48-54 to 
```go
indexes := []mgo.Index{
	{Key: []string{"-score"}},
	{Key: []string{"src", "fqdn", "src_network_uuid"}},
	{Key: []string{"src", "src_network_uuid"}},
	{Key: []string{"fqdn"}},
//	{Key: []string{"proxy.ip", "proxy.network_uuid"}}, this one's up to you if you think we might need it in the future
	{Key: []string{"-connection_count"}},
}
```

This incorporates the original index fix that was included in the first draft of #689.

fullmetalcache replied:
Changed the indexes to the ones you mentioned

Zalgo2462 commented on config/tables.go:
```diff
@@ -21,13 +21,14 @@ type (
+		OpenConnTable        string `default:"openconn"`
+		SSLTable             string `default:"ssl"`
+		UniqueConnTable      string `default:"uconn"`
+		UniqueConnProxyTable string `default:"uconnproxy"`
```
Could we use `uconnProxy` so that way we don't have to remember which of `beaconProxy` and `unconnProxy` is uppercase? It looks like most of the multiword collections added so far are camelCase. I don't care much if we go all lowercase or camelCase, but I would like to keep it consistent.

fullmetalcache replied:
Updated to camelCase for the collection name

Zalgo2462 commented on pkg/beaconproxy/analyzer.go:
```diff
@@ -65,42 +66,32 @@ func (a *analyzer) start() {
+			// if uconnproxy has turned into a strobe, we will not have any timestamps here,
+			// and we need to update uconnproxy table with the strobe flag. This is being done
+			// here and not in uconnproxy because uconnproxy doesn't do reads, and doesn't know
+			// the updated conn count
```
Are strobes ever turned back into regular unique connections if their connection counts decrement below the threshold? I think there might be a bug in the original beaconing logic regarding this as well.

Would a fix be to simply add "$set": bson.M{"strobe": false} to query when `TsList != nil` here and in the beacons package?

fullmetalcache replied:
Yep, that should do it! I added in the extra set field for the not-strobe case. Good catch! I updated in uconn and uconnproxy as well, since we set strobe fields there too.

Also, realized the strobe field for beaconproxy is stored as "strobe" from the RITA side and is attempted to be retrieved with strobeFQDN in the middleware. This should probably be strobeProxy in both places. I can't remember if there was a reason for not just naming it strobe. Thoughts?

In the uconnproxy collection, I am just going to keep it as "strobe" unless there are objections.

Zalgo2462 replied:
I'm good with whatever makes the least amount of work. I don't think there was a specific reason for calling it one thing over another. I'm also okay with using `strobeFQDN` in the beaconProxy collection in addition to the beaconFQDN collection. At the end of the day, both collections represent strobes going out to FQDNs. 

However, if you want to call it strobeProxy here and issue a fix-up to the middleware that works for me too.

fullmetalcache replied:
Cool, we can just leave it as strobeFQDN to reduce the workload

Zalgo2462 commented on pkg/uconnproxy/mongodb.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	indexes := []mgo.Index{
+		{Key: []string{"src", "src_network_uuid", "fqdn"}, Unique: true},
```
Please  swap the ordering of this key to be 
`{Key: []string{"src", "fqdn", "src_network_uuid"}, Unique: true},`

MongoDB can use the prefix of a compound key to service a query. By reordering this, we gain the ability to search for (src, fqdn) pairs without the network uuid which can be helpful for quick debugging. In the current ordering, we have the ability to search for (src, src_network_uuid), but that search is already covered by the key below.

fullmetalcache replied:
I was not aware of that, good to know! Re-ordered

Zalgo2462 reviewed (changes requested):
- Found a few small things to change, but generally they relate to simplifying the code or naming rather than correctness
- We may have a bug where rolling strobes are never turned back into normal connections when their connection counters go back down
- Found some performance items that would be easier to fix up now before merging with #654.

fullmetalcache commented on parser/fsimporter.go:
```diff
@@ -672,6 +677,31 @@ func (fs *FSImporter) parseFiles(indexedFiles []*fpt.IndexedFile, parsingThreads
+								// handle if URI is suffixed with a path like
+								// /index.html, /home, etc.
+								if strings.Contains(uri, "/") {
+									uri = strings.Split(uri, "/")[0]
```
Updated, tested, and added comments. Thanks for providing the logic! I do remember now you mentioning the speed difference between Split and Slice/Index. Will keep that in mind for the future.

fullmetalcache commented on parser/fsimporter.go:
```diff
@@ -714,25 +744,24 @@ func (fs *FSImporter) parseFiles(indexedFiles []*fpt.IndexedFile, parsingThreads
+									// create new uconnProxyMap record with src and fqdn
+									uconnProxyMap[srcFQDNKey] = &uconnproxy.Input{
+										Hosts: srcFQDNPair,
+										Proxy: dstUniqIP,
```
Sounds good!

fullmetalcache commented on pkg/beaconfqdn/repository.go:
```diff
@@ -48,36 +48,37 @@ type (
+		Ts             TSData           `bson:"ts"`
+		Ds             DSData           `bson:"ds"`
+		Score          float64          `bson:"score"`
+		ResolvedIPs    data.UniqueIPSet `bson:"resolved_ips"`
```
Changed it back

fullmetalcache commented on pkg/uconnproxy/repository.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+//NewUniqueSrcHostname binds a pair of UniqueIPs and an FQDN
+func NewUniqueSrcHostname(source data.UniqueIP, fqdn string) UniqueSrcHostname {
```
Moved the duplicated methods and functions into data/fqdn.go. I renamed uniqueSrcHostnamePair to be uniqueSrcFQDNPair as I think hostname could be a bit ambiguous as to whether we are talking about a system with a hostname or an fqdn that points to a system.

fullmetalcache commented on pkg/beaconproxy/mongodb.go:
```diff
@@ -125,15 +125,12 @@ func (r *repo) Upsert(proxyHostnameMap map[string]*Input) {
 
 		// progress bar increment
-		bar.IncrBy(1, time.Since(start))
+		bar.IncrBy(1)
```
Yes, I went ahead and updated it in all of the packages. Not sure what sorcery was allowing it to continue working with stuffing in the extra argument. My linter was complaining about it and is what led me to it. This is the function definition that I see in the package:

![image](https://user-images.githubusercontent.com/11990044/129936751-46612c4e-5428-498e-baa2-414fcbd0e91b.png)

Could be something weird with my system too?

Zalgo2462 replied:
![image](https://user-images.githubusercontent.com/761220/129970665-21ca0931-c057-416b-acb4-3f62e8cb7331.png)
This is the definition for the function in the mpb package on my machine. It says its held at version 3.3.4. I think you can set up the progress bars to give an estimated finish time, for which it needs the delta times. However, since we don't display anything like that, I think we're good to remove the delta times anyways.

Zalgo2462 reviewed (changes requested):
I was able to split up a dataset we had used for testing the proxy functionality into chunks using a Python script I whipped up. After doing so, the new PR seemed to work well. The results matched the results from importing all of the files at once, so the main issue has definitely been solved :tada:. 

However, old data is not being removed from the `beaconProxy` or `uconnProxy` collection when the chunk ID wraps back around. 

Please add any logic necessary for removing outdated entries to the `pkg/remover` package.
It may be as simple as adding the two collections to the `removeOutdatedCIDs()` method in `pkg/remover/mongodb.go`.

fullmetalcache commented:
Added the lines for removing old chunks and confirmed working. Split http.log in half, manually imported each file and specified the same CID for both import, confirmed that the chunks are no longer duplicated.

Zalgo2462 reviewed (approved):
LGTM
